File: dsa/binary_search/367_Valid_Perfect_Square.java
package binary_search;

class Solution {
    public boolean isPerfectSquare(int num) {
        int left = 1;
        int right = num;

        while (left <= right){
            int mid = left + (right-left)/2;
            if ((long) mid * mid == (long)num){
                return true;
            }
            else if ((long)mid * mid > (long)num){
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }
        return false;
    }
}

File: dsa/5.binary_search/367_Valid_Perfect_Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        left: int = 1
        right: int = num

        while left <= right:
            mid: int = int((right - left) / 2) + left
            pow2_val: int = mid * mid
            if pow2_val == num:
                return True
            elif pow2_val < num:
                left = mid + 1
            else:
                right = mid - 1
        
        return False
